blockindex/liquidstaking_indexer.go

```diff
@@ -0,0 +1,958 @@
+	}
+	newBtIdx, ok := dirty.getBucketTypeIndex(amountParam, bt.Duration)
+	if !ok {
+		return errors.Wrapf(errBucketTypeNotExist, "amount %d, duration %d", amountParam.Int64(), bt.Duration)
+	}
+	b.TypeIndex = newBtIdx
+	return dirty.updateBucketInfo(tokenIDParam.Uint64(), b)
+}
+
+func (s *liquidStakingIndexer) handleDelegateChangedEvent(dirty *liquidStakingDirty, event eventParam) error {
+	tokenIDParam, err := event.indexedFieldUint256("tokenId")
+	if err != nil {
+		return err
+	}
+	delegateParam, err := event.fieldAddress("newDelegate")
+	if err != nil {
+		return err
+	}
+
+	b, ok := dirty.getBucketInfo(tokenIDParam.Uint64())
+	if !ok {
+		return errors.Wrapf(ErrBucketInfoNotExist, "token id %d", tokenIDParam.Uint64())
+	}
+	b.Delegate = delegateParam
+	return dirty.updateBucketInfo(tokenIDParam.Uint64(), b)
+}
+
+func (s *liquidStakingIndexer) handleWithdrawalEvent(dirty *liquidStakingDirty, event eventParam) error {
+	tokenIDParam, err := event.indexedFieldUint256("tokenId")
+	if err != nil {
+		return err
+	}
+
+	return dirty.burnBucket(tokenIDParam.Uint64())
+}
+
+func (s *liquidStakingIndexer) loadCache() error {
+	delta := newLiquidStakingDelta()
+	// load height
+	var height uint64
+	h, err := s.kvstore.Get(_liquidStakingNS, _liquidStakingHeightKey)
+	if err != nil {
+		if !errors.Is(err, db.ErrNotExist) {
+			return err
+		}
+		height = 0
+	} else {
+		height = byteutil.BytesToUint64BigEndian(h)
+
+	}
+	delta.putHeight(height)
+
+	// load total bucket count
+	var totalBucketCount uint64
+	tbc, err := s.kvstore.Get(_liquidStakingNS, _liquidStakingTotalBucketCountKey)
+	if err != nil {
+		if !errors.Is(err, db.ErrNotExist) {
+			return err
+		}
+	} else {
+		totalBucketCount = byteutil.BytesToUint64BigEndian(tbc)
+	}
+	delta.putTotalBucketCount(totalBucketCount)
+
+	// load bucket info
+	ks, vs, err := s.kvstore.Filter(_liquidStakingBucketInfoNS, func(k, v []byte) bool { return true }, nil, nil)
+	if err != nil {
+		if !errors.Is(err, db.ErrBucketNotExist) {
+			return err
+		}
+	}
+	for i := range vs {
+		var b BucketInfo
+		if err := b.deserialize(vs[i]); err != nil {
+			return err
+		}
+		delta.putBucketInfo(byteutil.BytesToUint64BigEndian(ks[i]), &b)
+	}
+
+	// load bucket type
+	ks, vs, err = s.kvstore.Filter(_liquidStakingBucketTypeNS, func(k, v []byte) bool { return true }, nil, nil)
+	if err != nil {
+		if !errors.Is(err, db.ErrBucketNotExist) {
+			return err
+		}
+	}
+	for i := range vs {
+		var b BucketType
+		if err := b.deserialize(vs[i]); err != nil {
+			return err
+		}
+		delta.putBucketType(byteutil.BytesToUint64BigEndian(ks[i]), &b)
+	}
+	return s.cache.merge(delta)
+}
+
+func (s *liquidStakingIndexer) commit(dirty *liquidStakingDirty) error {
+	batch, delta := dirty.finalize()
+	if err := s.kvstore.WriteBatch(batch); err != nil {
+		return err
+	}
+	if err := s.cache.merge(delta); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (s *liquidStakingIndexer) blockHeightToDuration(height uint64) time.Duration {
```

Comment:
as we discussed, we shouldn't use `time.Duration`

Reply:
a strange func for me, and it isn't used.

Reply:
a deprecated method, and should be removed